File: Algo/Leetcode/207CourseSchedule.cpp

class Solution {
public:
    vector<vector<int>> edges;
    vector<int> visit;
    bool valid = true;

    void dfs(int i){
        visit[i] = 1;
        for(auto e:edges[i]){
            if(visit[e]==0){
                dfs(e);
                if(!valid) return;
            } else if(visit[e]==1){
                valid=false;
                return;
            }
        }
        visit[i] = 2;
    }

    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        for(int i=0; i<numCourses; i++){
            edges.push_back({});
            visit.push_back(0);
        }

        for(auto c:prerequisites) edges[c[1]].push_back(c[0]);

        for(int i=0; i<numCourses; i++){
            if(visit[i]==0) dfs(i);
            if(!valid) return false;
        }
        return true;
    }
};

File: Algo/Leetcode/207CourseSchedule.py
class Solution(object):

    def __init__(self):
        self.valid = True
        self.visit = None
        self.edge = None

    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        n = numCourses
        self.visit = [0] * n
        self.edge = [[] for _ in range(n)]
        for p in prerequisites:
            self.edge[p[1]].append(p[0])

        for i in range(n):
            if self.visit[i] == 0:
                self.dfs(i)
            if not self.valid:
                return False

        return True

    def dfs(self, i):
        self.visit[i] = 1
        for e in self.edge[i]:
            if self.visit[e] == 0:  # not loop, dfs
                self.dfs(e)
                if not self.valid:      # early stop
                    return
            elif self.visit[e] == 1:
                self.valid = False
                return

        self.visit[i] = 2